Develop

yapracticumplakhotnik opened the pull request.

Commit:
add allure report

Commit:
Финальный проект спринт 7

pomegranate282922 commented on src/test/java/tests/BaseTest.java:
```diff
@@ -0,0 +1,20 @@
+        RestAssured.baseURI = baseUrl;
+    }
+
+    protected RequestSpecification givenRequest() {
```
Отлично что применяется req spec

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Step("Метод POST to /api/v1/courier - курьера можно создать")
+    public void createCourier() {
+        // Создаем курьера перед каждым тестом
+        String body = "{ \"login\": \"" + courierLogin + "\", \"password\": \"" + courierPassword + "\", \"firstName\": \"" + courierName + "\" }";
```
Тело запроса передаем через объект 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,86 @@
+        // Создаем курьера перед каждым тестом
+        String body = "{ \"login\": \"" + courierLogin + "\", \"password\": \"" + courierPassword + "\", \"firstName\": \"" + courierName + "\" }";
+
+        givenRequest()
```
Все вызовы ручек выносим в отдельные классы и переиспользуем 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,86 @@
+        givenRequest()
+                .body(body)
+                .when()
+                .post("/api/v1/courier")
```
Все урлы выносим как константы 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    @Test
+    @Step("Метод POST to /api/v1/courier - запрос возвращает правильный код ответа и успешный запрос возвращает ok: true")
```
Тест это не шаг. Здесь используем аннотацию DisplayName 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,86 @@
+    @After
+    public void tearDown() {
+        if (courierId != 0) {
+           deleteCourier(courierId);
```
Удаление не происходит( 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierLoginTest.java:
```diff
@@ -0,0 +1,69 @@
+
+    @After
+    public void deleteCourier() {
+        //  удалить курьера по логину
```
? Не доделано 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/OrderCreationTest.java:
```diff
@@ -0,0 +1,59 @@
+    @Parameterized.Parameters
+    public static Collection<Object[]> colors() {
+        return Arrays.asList(new Object[][]{
+                {null}, // без цвета
```
Отлично применена параметризация

pomegranate282922 reviewed (commented):
Работа хорошая! Все тесты реализованы. Применена параметризация. 
замечания:
1. Все вызовы ручек выносим из тестовых классов 
2. Все урлы выносим как константы 
3. Поправим аннотации к тестам 
4. Все тела запросов передаем через объект 
5. Разобраться с удалением тестовых данных. Все созданные курьеры нужно удалить.

Commit:
Исправлены:
аннотации
удаление курьера
вынесены в константы урлы и ручки
тело запроса через объект

pomegranate282922 commented on src/test/java/tests/CourierLoginTest.java:
```diff
@@ -0,0 +1,71 @@
+    @Before
+    public void createCourier() {
+        String body = "{ \"login\": \"" + login + "\", \"password\": \"" + password + "\", \"firstName\": \"AuthName\" }";
+        givenRequest()
```
Все(!) вызовы ручек выносим из тестовых классов 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,113 @@
+    public String courierPassword = "\"password\": \" + testPassword + \"";
+    public   String courierName = "\"firstName\": \" + testName + \"}";
+
+    static class CreateCourier {
```
Это делаем отдельным классом и везде переиспользуем 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,113 @@
+    @DisplayName("Метод POST to /api/v1/courier - курьера можно создать")
+    public void createCourier() {
+        // Создаем курьера перед каждым тестом
+        CreateCourier createCourierBody = new CreateCourier(courierLogin,courierPassword,courierName);
```
В параметры надо передать именно значения полей, а не строку json 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,113 @@
+    @Test
+    @DisplayName("Удаление курьера")
+    public void delete(){
+        tearDown();
```
Удаление курьера НЕ происходит. Прошу разобраться почему и исправить ошибку 
Нужно исправить

pomegranate282922 reviewed (commented):
Привет! Многое поправил, но к сожалению не все. 
все замечания нужно применить ко всему коду.

Commit:
Исправлены:
удаление расписала уже чтобы наверняка
постарась в параметры передать значения полей, а не строку json
вынесла ручки

pomegranate282922 reviewed (commented):
Стало лучше - супер! Однако не исправлены важные замечания:
1. Все(!!!) вызовы ручек нужно вынести из тестового класса и переиспользовать. Сделать методы, возвращающие Response - вызываем их в тестах и анализируем ответ в зависимости от ситуации.
2. Тело запроса передаем через объект - нужно создать нужного класса. не через строку. 
3. Каждый тест после себя должен удалять созданного курьера. В классе на создание курьера этого не делается 
4. Все внутренние статик классы сделать отдельными классами

pomegranate282922 commented on src/test/java/tests/BaseTest.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    static class CreateCourier {
+        String login;
```
Все эти статик классы выносим в отдельные классы

pomegranate282922 commented on src/test/java/tests/BaseTest.java:
```diff
@@ -0,0 +1,121 @@
+    }
+
+
+    @DisplayName("Метод POST to /api/v1/courier - создать курьера")
```
Этой аннотацией помечаем только тесты. Это тест? Нет. Это шаг.

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,78 @@
+    @Test
+    @DisplayName("Метод POST to /api/v1/courier - запрос возвращает правильный код ответа и успешный запрос возвращает ok: true")
+    public void testCreateCourierSuccess() {
+        givenRequest()
```
Вызовы ручек выносим в отдельные классы - делаем там метод с вызовом, возвращающим Response. В параметре - объект который надо передать в тело запроса. 
Здесь( в тестах) вызываем этот метод, получаем респонс из него и анализируем результат - статус и тело ответа. 
такое нужно применить ко всем (!) вызовам ручек во всем коде.

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,78 @@
+                .then()
+                .statusCode(201)
+                .body("ok", equalTo(true));
+
```
Здесь ты создала курьера. Теперь ( после теста) его нужно удалить. То есть в метод с аннотацией after добавить удаление. 
аналогично нужно чтобы после каждого теста, который создает курьера, он удалялся

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,78 @@
+                .when()
+                .post(createCourier)
+                .then()
+                .statusCode(409); // ожидается конфликт
```
Вот здесь тебе нужно сделать дублирование. Для этого тест должен - сначала создать один курьер, потом попытаться второй. В конце удалить созданного первого курьера,

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,78 @@
+    @DisplayName("Метод POST to /api/v1/courier - если одного из полей нет, запрос возвращает ошибку")
+    public void testCreateCourierMissingFields() {
+
+        String body = "{ \"login\": \"" + System.currentTimeMillis() + "\" }";
```
Тело запроса это не должна дать строка, это должен дать объект нужного класса ( например логин или create или тп). Используем сериализацию встроенную в RestAssured

pomegranate282922 commented on src/test/java/tests/CourierLoginTest.java:
```diff
@@ -0,0 +1,80 @@
+
+public class CourierLoginTest extends BaseTest {
+
+    @Test
```
Это тест? Нет.

pomegranate282922 commented on src/test/java/tests/OrderCreationTest.java:
```diff
@@ -0,0 +1,60 @@
+    @Test
+    @DisplayName("Метод POST to /api/v1/orders - создаём заказ используя параметризацию")
+    public void testCreateOrder() {
+        String body = "{"
```
То же самое - передать нужно объект - с такими полями,

pomegranate282922 reviewed (commented):
Привет, написала подробные комментарии. Их нужно применить ко всему коду. 
и обязательно разобраться с аннотациями before и after и для чего они нужны - и применить их

Commit:
по совету наставника использовала его объяснения с вебинара

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,57 @@
+        Courier courier = new Courier("testCourierLoginJenya","testPassword","testName");
+        ValidatableResponse response = client.createCourier(courier);
+        response.assertThat().statusCode(201).and().body("ok",equalTo(true));
+    }
```
После этого теста должно дать удаление созданного курьера 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,57 @@
+    @Test
+    @DisplayName("Метод POST to /api/v1/courier - нельзя создать двух одинаковых курьеров - получим 409 код")
+    public void test2CreateDuplicateCourier() {
+        QAScooterAPIClient client = new QAScooterAPIClient();
```
Здесь должно быть: в начале создаем первого курьера, потом попытка создать  дополнительного, в конце удаление первого 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierCreationTest.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Test
+    @DisplayName("Метод POST to /api/v1/courier - если одного из полей нет, запрос возвращает ошибку")
+    public void test3CreateCourierMissingFields() {
```
Магические цифры в именах не используем. Тесты должны быть абсолютно независимыми от порядка исполнения 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/CourierLoginTest.java:
```diff
@@ -0,0 +1,58 @@
+        int statusCode = response.extract().statusCode();
+        Assertions.assertEquals(400, statusCode);
+
+    }
```
Курьер создается перед каждым тестом. Но нигде нет удаления. 
нужно исправить

pomegranate282922 reviewed (commented):
Привет! Стало намного лучше, супер! Но остались важные вопросы, с которыми необходимо разобраться. Прошу еще додумать.

Commit:
удаление и цифры поправлены

pomegranate282922 reviewed (approved):
Ок, зачет.
Но - советую еще раз разобраться с применением аннотаций before и after, сейчас очень много дублирования кода. 
также не хардкодим тестовые данные в методах -шагах. Передаем их через параметр.
